app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,207 @@
+from django.test import TestCase
+from django.urls import reverse
+
+from rest_framework.test import APIClient
+from rest_framework import status
+
+from core.models import Recipe, Ingredient
+
+RECIPE_URL = reverse('recipe:recipe-list')
+
+
+def recipe_url(id):
+    """Construct URL for a single recipe based on its ID"""
+    return reverse('recipe:recipe-detail', args=[id])
+
+
+def create_sample_recipe(**params):
+    """Helper function to create a user"""
+    return Recipe.objects.create(**params)
+
+
+class RecipeAPITests(TestCase):
+
+    def setUp(self):
+        self.client = APIClient()
+
+    def test_create_recipe_with_ingredients(self):
+        """Test creating a recipe including ingredients"""
+        payload = {
+            'name': 'Vegan Roast Dinner',
+            'description': 'Roasted potatoes and mushroom wellington'
+                           ' with vegetables and gravy.',
+            'ingredients': [
+                {'name': 'carrots'},
+                {'name': 'potatoes'},
+                {'name': 'mushrooms'},
+            ]
+        }
+
+        response = self.client.post(RECIPE_URL, payload, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(
+            payload['name'],
+            Recipe.objects.get(id=response.data['id']).name
+        )
+        self.assertEquals(
+            len(response.data['ingredients']),
+            len(payload['ingredients'])
+        )
+
+    def test_get_recipes(self):
+        """Test retrieving a recipe"""
+        create_sample_recipe(
+            name='Roast Dinner',
+            description='Roasted potatoes and chicken'
+                        ' with vegetables and gravy.'
+        )
+
+        create_sample_recipe(
+            name='Beans on Toast',
+            description='Just the best.'
+        )
+
+        response = self.client.get(RECIPE_URL)
+
+        recipes = Recipe.objects.all().order_by('-name')
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), len(recipes))
+
+    def test_get_recipe(self):
+        """Test retrieving a single recipe using name as filter"""
+        test_recipe_name = 'Beans on Toast'
+
+        create_sample_recipe(
+            name='Roast Dinner',
+            description='Roasted potatoes and chicken'
+                        ' with vegetables and gravy.'
+        )
+
+        create_sample_recipe(
+            name=test_recipe_name,
+            description='Just the best recipe.'
+        )
+
+        response = self.client.get(RECIPE_URL, {'name': test_recipe_name})
+
+        recipes = Recipe.objects.all().order_by('-name')
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertNotEqual(len(response.data), len(recipes))
```

Comment:
Same as above, no real improvement here, but better than `len(recipes)`, which retrieves all recipe columns (`SELECT * FROM recipe`) you could just use the queryset and do a `COUNT`
```suggestion
        self.assertNotEqual(len(response.data), recipes.count())
```